File: Python/libraries/recognizers-text/recognizers_text/culture.py
#  Copyright (c) Microsoft Corporation. All rights reserved.
#  Licensed under the MIT License.

class Culture:
    Chinese: str = 'zh-cn'
    Dutch: str = 'nl-nl'
    English: str = 'en-us'
    EnglishOthers: str = 'en-*'
    French: str = 'fr-fr'
    Italian: str = 'it-it'
    Japanese: str = 'ja-jp'
    Korean: str = 'ko-kr'
    Portuguese: str = 'pt-br'
    Spanish: str = 'es-es'
    SpanishMexican: str = 'es-mx'
    Turkish: str = 'tr-tr'
    German: str = 'de-de'

    @staticmethod
    def _get_supported_culture_codes():
        return [
            Culture.English,
            Culture.EnglishOthers,
            Culture.Dutch,
            Culture.Chinese,
            Culture.French,
            Culture.Italian,
            Culture.Japanese,
            Culture.Korean,
            Culture.Portuguese,
            Culture.Spanish,
            Culture.SpanishMexican,
            Culture.Turkish,
            Culture.German
        ]

    @staticmethod
    def map_to_nearest_language(culture_code: str):
        if not culture_code:
            return
        culture_code = culture_code.lower()
        supported_culture_codes = Culture._get_supported_culture_codes()
        if culture_code not in supported_culture_codes:
            culture_prefix = culture_code.split('-')[0].strip()
            possible_cultures = []
            for supportedCultureCode in supported_culture_codes:
                if supportedCultureCode.startswith(culture_prefix):
                    possible_cultures.append(supportedCultureCode)

            if possible_cultures:
                if len(possible_cultures) > 1:
                    for code in possible_cultures:
                        if '*' in code:
                            culture_code = code
                else:
                    culture_code = possible_cultures[0]
        return culture_code


class BaseCultureInfo:
    def __init__(self, culture_code: str):
        self.code: str = culture_code

    def format(self, value: object) -> str:
        return repr(value)

File: JavaScript/packages/recognizers-text/src/culture.ts
export class Culture {

  static readonly English: string = "en-us"
  static readonly EnglishOthers: string = "en-*"
  static readonly Chinese: string = "zh-cn"
  static readonly Spanish: string = "es-es"
  static readonly Portuguese: string = "pt-br"
  static readonly French: string = "fr-fr"
  static readonly German: string = "de-de"
  static readonly Japanese: string = "ja-jp"
  static readonly Dutch: string = "nl-nl"
  static readonly Italian: string = "it-it"

  static readonly supportedCultures: Culture[] = [
    new Culture("English", Culture.English),
    new Culture("EnglishOthers", Culture.EnglishOthers),
    new Culture("Chinese", Culture.Chinese),
    new Culture("Spanish", Culture.Spanish),
    new Culture("Portuguese", Culture.Portuguese),
    new Culture("French", Culture.French),
    new Culture("German", Culture.German),
    new Culture("Japanese", Culture.Japanese),
    new Culture("Dutch", Culture.Dutch),
    new Culture("Italian", Culture.Italian)
  ]

  readonly cultureName: string
  readonly cultureCode: string

  protected constructor(cultureName: string, cultureCode: string) {
    this.cultureName = cultureName;
    this.cultureCode = cultureCode;
  }

  static getSupportedCultureCodes(): string[] {
    return Culture.supportedCultures.map(c => c.cultureCode);
  }

  static mapToNearestLanguage(cultureCode: string): string {
    if (cultureCode !== undefined) {
      cultureCode = cultureCode.toLowerCase();
      let supportedCultureCodes = Culture.getSupportedCultureCodes();

      if (supportedCultureCodes.indexOf(cultureCode) < 0) {
        let culturePrefix = cultureCode.split('-')[0].trim();

        supportedCultureCodes.forEach(function (supportedCultureCode) {
          if (supportedCultureCode.startsWith(culturePrefix)) {
            cultureCode = supportedCultureCode;
          }
        });
      }
    }

    return cultureCode;
  }
}

export class CultureInfo {
  readonly code: string;

  static getCultureInfo(cultureCode: string): CultureInfo {
    return new CultureInfo(cultureCode);
  }

  constructor(cultureName: string) {
    this.code = cultureName;
  }
}